python/src/iceberg/literals.py

```diff
@@ -0,0 +1,595 @@
+    EPOCH_DAY = EPOCH.date()
+
+    @staticmethod
+    def above_max():
+        return ABOVE_MAX
+
+    @staticmethod
+    def below_min():
+        return BELOW_MIN
+
+
+def from_(value):  # noqa: C901
+    if value is None:
+        raise RuntimeError("Cannot create an expression literal from None")
+    if isinstance(value, bool):
+        return BooleanLiteral(value)
+    elif isinstance(value, int):
+        if Literal.JAVA_MIN_INT < value < Literal.JAVA_MAX_INT:
+            return IntegerLiteral(value)
+        return LongLiteral(value)
+    elif isinstance(value, float):
+        if Literal.JAVA_MIN_FLOAT < value < Literal.JAVA_MAX_FLOAT:
+            return FloatLiteral(value)
+        return DoubleLiteral(value)
+    elif isinstance(value, str):
+        return StringLiteral(value)
+    elif isinstance(value, uuid.UUID):
+        return UUIDLiteral(value)
+    elif isinstance(value, bytearray):
+        return BinaryLiteral(value)
+    elif isinstance(value, bytes):
+        return FixedLiteral(value)
+    elif isinstance(value, Decimal):
+        return DecimalLiteral(value)
+    else:
+        raise NotImplementedError("Unimplemented Type Literal for value: %s" % value)
+
+
+class Literal(object):
+    JAVA_MAX_INT = 2147483647
+    JAVA_MIN_INT = -2147483648
+    JAVA_MAX_FLOAT = 3.4028235e38
+    JAVA_MIN_FLOAT = -3.4028235e38
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def to_byte_buffer(self):
+        raise NotImplementedError()
+
+
+def of(value):  # noqa: C901
+    if isinstance(value, bool):
+        return BooleanLiteral(value)
+    elif isinstance(value, int):
+        if value < Literal.JAVA_MIN_INT or value > Literal.JAVA_MAX_INT:
+            return LongLiteral(value)
+        return IntegerLiteral(value)
+    elif isinstance(value, float):
+        if value < Literal.JAVA_MIN_FLOAT or value > Literal.JAVA_MAX_FLOAT:
+            return DoubleLiteral(value)
+        return FloatLiteral(value)
+    elif isinstance(value, str):
+        return StringLiteral(value)
+    elif isinstance(value, uuid.UUID):
+        return UUIDLiteral(value)
+    elif isinstance(value, bytes):
+        return FixedLiteral(value)
+    elif isinstance(value, bytearray):
+        return BinaryLiteral(value)
+    elif isinstance(value, Decimal):
+        return DecimalLiteral(value)
+
+
+class BaseLiteral(Literal):
+    """Base literal for all"""
+
+    def __init__(self, value):
+        self.value = value
+        self.byte_buffer = None
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def __eq__(self, other):
+        if id(self) == id(other):
+            return True
+        elif other is None or not isinstance(other, BaseLiteral):
+            return False
+
+        return self.value == other.value
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        return "BaseLiteral(%s)" % str(self.value)
```

Comment:
Is it true that we don't expect BaseLiteral to ever be instantiated directly? If so maybe we should remove this `__repr__` and instead add `__repr__` methods directly to the subclasses like `BooleanLiteral` and `IntegerLiteral`. Or you can take the same approach as [types.py](https://github.com/apache/iceberg/blob/master/python/src/iceberg/types.py#L49-#L58) where the repr and str methods are defined as part of the super init.